homeassistant/components/sonos/speaker.py

```diff
@@ -217,3 +244,57 @@ async def async_unseen(self, now: datetime.datetime | None = None) -> None:
             await subscription.unsubscribe()
 
         self._subscriptions = []
+
+    async def async_update_device_properties(self, event: SonosEvent = None) -> None:
+        """Update device properties using the provided SonosEvent."""
+        if event is None:
+            return
+
+        if (more_info := event.variables.get("more_info")) is not None:
+            battery_dict = dict(x.split(":") for x in more_info.split(","))
+            await self.async_update_battery_info(battery_dict)
+
+        self.async_write_entity_states()
+
+    async def async_update_battery_info(self, battery_dict: dict[str, Any]) -> None:
+        """Update battery info using the decoded SonosEvent."""
+        self._last_battery_event = dt_util.utcnow()
+
+        is_charging = EVENT_CHARGING[battery_dict["BattChg"]]
+        if is_charging == self.charging:
+            self.battery_info.update({"Level": int(battery_dict["BattPct"])})
+        else:
+            if battery_info := await self.hass.async_add_executor_job(
+                fetch_battery_info_or_none, self.soco
+            ):
+                self.battery_info = battery_info
+
+    @property
+    def power_source(self) -> str:
+        """Return the name of the current power source.
+
+        Observed to be either BATTERY or SONOS_CHARGING_RING or USB_POWER.
+        """
+        return self.battery_info.get("PowerSource", STATE_UNKNOWN)
+
+    @property
+    def charging(self) -> bool:
+        """Return the charging status of the speaker."""
+        return self.power_source not in ("BATTERY", STATE_UNKNOWN)
+
+    async def async_poll_battery(self, now: datetime.datetime | None = None) -> None:
+        """Poll the device for the current battery state."""
+        if not self.available:
+            return
+
+        if (
+            self._last_battery_event
+            and dt_util.utcnow() - self._last_battery_event < BATTERY_SCAN_INTERVAL
+        ):
+            return
+
+        if battery_info := await self.hass.async_add_executor_job(
+            fetch_battery_info_or_none, self.soco
+        ):
+            self.battery_info = battery_info
+            self.async_write_entity_states()
```

Comment:
Is this going to update entities that have no changes?

Reply:
As of now, yes.